Add HeaderDecoder with custom header name matching

paulpdaniels opened the pull request:
Closes #315 

Adds a new set of `decoderWith` operators to the `HeaderDecoder` and `HeaderCodec` companion objects. The old method `decoder` method now delegates to `decoderWith` but uses the default equality check for strings.

This allows users to override how the strings are compared when trying to determine the order of the headers in the mapping, e.g. when comparing the strings in a case insensitive manner.

nrinaudo commented:
Right, this is really something I should put in the contribution guidelines. There's no issue that I can see with your PR, aside from a purely cosmetic one: the code is not formatted according to the project's guidelines.

You just need to run the `scalafmtAll` SBT task and it'll fix the problem(s).

nrinaudo commented on core/shared/src/main/boilerplate/kantan/csv/GeneratedHeaderDecoders.scala.template:
```diff
@@ -31,11 +31,28 @@ trait GeneratedHeaderDecoders {
+   *
+   * scala> "[#f1#]\n[#1#]".asCsvReader[Foo](rfc.withHeader).next()
+   * res0: ReadResult[Foo] = Right(Foo([#1#,]))
+   * }}
```
This should be `}}}`, not `}}`:

```
[warn] /path/to/kantan/csv/core/jvm/target/scala-2.12/src_managed/main/kantan/csv/GeneratedHeaderDecoders.scala:754:3: unclosed code block
[warn]   /** Creates a new [[HeaderDecoder]] from the specified field list, and an equality function, and function to convert the fields into the resulting target type.
[warn]   ^
```

nrinaudo commented:
Also: given that your use case is likely quite a common one, I wouldn't be against pre-defined helpers. We have one for case-sensitive headers, why not:

```scala
  [#
  def caseInsensitiveDecoder[[#A1: CellDecoder#], R]([#f1: String#])(f: ([#A1#]) => R): HeaderDecoder[R] =
    decoderWith([#f1#])(_.equalsIgnoreCase(_))(f)#
  ]
```

paulpdaniels commented:
@nrinaudo comments addressed

nrinaudo commented:
I, err... were they? I'm not as used to the Github UI as I should be, but checking out the latest version of the branch locally still has `}}` rather than `}}}` in `GeneratedHeaderDecoders`, which triggers compile time warnings (and probably disable some doctests, although I haven't checked).

paulpdaniels commented:
Ah ok, I switched computers because scalafmt on my windows was going crazy, so I must have lost that change. Will fix

paulpdaniels commented:
Ok I think they should be fixed now 🤞

paulpdaniels commented:
Apologies I lost track of this and #319 until I got the notification this morning. Anyway I think I addressed your comments @nrinaudo let me know if there is anything else you want on this.
